Fix/gh issue 5797

hannes-steffenhagen-diffblue opened the pull request:
This fixes a small logic issue in `unpack_rec`. See https://github.com/diffblue/cbmc/issues/5797 for details.


- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Refactor unpack_rec for clarity

* Rename loop variables to make it more obvious what they represent
* Add some braces to if/else

Commit:
Fix bug in unpack_rec

The bug here was that a recent change to respect the max_bytes setting for
primitive types sometimes led to casts to too-small bitvectors because the same
variable was used for the "last" bit and the total amount of bits of the
primitive value.

See https://github.com/diffblue/cbmc/issues/5797

Co-authored-by: Michael Tauschnig <tautschn@amazon.com>
Co-authored-by: Thomas Given-Wilson <thomas.givenwilson@diffblue.com>

tautschnig commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -918,7 +919,7 @@ static exprt unpack_rec(
       extractbits_exprt extractbits(
         typecast_exprt::conditional_cast(
-          src, bv_typet{numeric_cast_v<std::size_t>(last_bit)}),
+          src, bv_typet{numeric_cast_v<std::size_t>(total_bits)}),
```
How about factoring this out of the loop completely? It's the same expression for each iteration of the loop, so only constructing it once would actually be beneficial.

hannes-steffenhagen-diffblue replied:
@tautschnig I can pull that change into the first refactoring commit, that would probably be an improvement yes.

tautschnig replied:
Just make sure you don't hoist it too high up so that you incidentally fix the issue by mere refactoring :-)

tautschnig reviewed (approved):
Thank you for the additional cleanup! I'm not entirely sure, but should the commit message actually say "Fixes: #5797" instead of "See #5797" to automatically close the issue once this PR is merged?

Commit:
Extract common expressions from loop in unpack_rec

These are going to repeated a couple of times for some larger types, there's no
need to recreate them over and over.

NlightNFotis reviewed (approved):
LGTM
